excel/src/main/java/org/apache/metamodel/excel/DefaultSpreadsheetReaderDelegate.java

```diff
@@ -167,50 +166,114 @@ private MutableTable createTable(final Workbook wb, final Sheet sheet) {
+                columnTypes[index] = ColumnType.STRING;
+            }
+        } else {
+            columnTypes[index] = columnType;
+        }
+        return columnTypes;
+    }
+
+    private void determineColumnDatatype(Object[] datatypes, Row row) {
+        for (int index = 0; index < row.getLastCellNum(); index++) {
+            CellType type = ((Cell) row.getCell(index)).getCellType();
+
+            if (datatypes[index] instanceof Object) {
+                datatypes[index] = type;
+            } else if (datatypes[index] instanceof CellType) {
+                if (datatypes[index].equals(type)) {
+                    continue;
+                } else {
+                    datatypes[index] = CellType.STRING;
+                }
+            }
+        }
+    }
+
     /**
      * Builds columns based on row/cell values.
      * 
      * @param table
      * @param wb
      * @param row
      */
-    private void createColumns(MutableTable table, Workbook wb, Row row) {
+    private void createColumns(MutableTable table, Workbook wb, Row row, ColumnType[] columTypes) {
         if (row == null) {
             logger.warn("Cannot create columns based on null row!");
             return;
         }
         final short rowLength = row.getLastCellNum();
-
         final int offset = getColumnOffset(row);
-
         // build columns based on cell values.
-        try (final ColumnNamingSession columnNamingSession = _configuration.getColumnNamingStrategy()
+        try (final ColumnNamingSession columnNamingSession = _configuration
+                .getColumnNamingStrategy()
                 .startColumnNamingSession()) {
             for (int j = offset; j < rowLength; j++) {
                 final Cell cell = row.getCell(j);
                 final String intrinsicColumnName = ExcelUtils.getCellValue(wb, cell);
                 final ColumnNamingContext columnNamingContext = new ColumnNamingContextImpl(table, intrinsicColumnName,
                         j);
                 final String columnName = columnNamingSession.getNextColumnName(columnNamingContext);
-                final Column column = new MutableColumn(columnName, ColumnType.VARCHAR, table, j, true);
+                Column column = null;
+                if (columTypes == null) {
+                    column = new MutableColumn(columnName, ColumnType.VARCHAR, table, j, true);
+                } else {
+                    column = new MutableColumn(columnName, columTypes[j], table, j, true);
+                }
                 table.addColumn(column);
             }
         }
     }
 
     /**
-     * Gets the column offset (first column to include). This is dependent on
-     * the row used for column processing and whether the skip empty columns
-     * property is set.
+     * Builds columns based on row/cell values.
+     * 
+     * @param table
+     * @param wb
+     * @param row
+     */
+    private void createColumns(MutableTable table, Workbook wb, Row row) {
```

Comment:
This method is now no longer used and as such can be removed.

Reply:
done with next commit